cache ipns entries to speed things up a little

whyrusleeping opened the pull request:
This adds a quick minute long cache to ipns entry resolution. 
Currently, ipns is pretty much unusable for webpages as each sub-request must re-resolve the entry.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

ion1 commented:
It would be good for to reflect that in a Cache-Control max-age header for /ipns addresses on the gateway.

kyledrake commented:
:+1:

whyrusleeping commented:
After this merges i encourage people to really start using ipns, and reporting *any* issues to me. I'm pretty confident in it (post this changeset), but thats not to say it won't have any issues.

ion1 commented:
I did an experiment with the current `master`, publishing a new object every few seconds and polling the IPNS address on the public gateway. It seemed to work impressively quickly at times while being slow to publish and resolve at other times.

I have been thinking of doing an experiment with live video streaming in this way. Unfortunately, this change might affect the current update rate negatively.

On the other hand, if I were to publish a blog using IPNS, I might be fine with nodes having a stale version of the record for up to an hour.

I realize future pubsub support will cover the live streaming use case, but having the ability to do `ipfs name publish --ttl=0` for the streaming experiment and `ipfs name publish --ttl=1h` for the blog example would let the publisher make the choice between low latency and greater caching (in this example, choosing to have a cache time an order of magnitude greater than the one minute).

Don’t let this delay merging this PR, but I hope IPNS records will gain a TTL field (like in DNS records) in the future.

lgierth commented on core/core.go:
```diff
@@ -238,6 +243,23 @@ func (n *IpfsNode) startOnlineServicesWithHost(ctx context.Context, host p2phost
+		return 0, err
+	}
+
+	ct := cfg.Ipns.ResolveCacheTime
```
Should we have this in the default config too?

whyrusleeping replied:
we can, but we dont need to. It assumes that an empty value means the default value of one minute.

lgierth replied:
i thought for the sake of documentation ;)

whyrusleeping replied:
added to the default config

jbenet commented on namesys/routing.go:
```diff
@@ -98,10 +130,35 @@ func (r *routingResolver) resolveOnce(ctx context.Context, name string) (path.Pa
+		eol, ok := checkEOL(entry)
+		if ok && eol.Before(cacheTil) {
+			cacheTil = eol
+		}
```
this stuff could get easily wrong, and cache time errors could behavior drastically. i'd propose feeding `entry` into `r.cacheSet` and put the computation in one place, calculated always from the original data, instead of intermediate values.

jbenet commented on routing/dht/routing.go:
```diff
@@ -84,7 +84,7 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key key.Key) ([]byte, error) {
 	defer cancel()
 
-	vals, err := dht.GetValues(ctx, key, 16)
+	vals, err := dht.GetValues(ctx, key, (KValue/2)+1)
```
- why `KValue/2` ? 
- document choice.

again, in a large network, this is totally going to be overshadowed by tons of records floating around. need a large sample to have hope of getting the latest.

whyrusleeping replied:
i performed a seance and the ghost of alan turing said that (K/2)+1 was a good value.

jbenet commented:
i'd support the addition of the --ttl flag for `name publish`, as a means to experiment for now (mark it experimental)

jbenet commented:
:+1: to this PR in general.

whyrusleeping commented:
@ion1 @jbenet i have a `--nocache` option that will ignore the cached value when requesting a record. I will also add the `--ttl`, but wiring that down is a little more intrusive.

whyrusleeping commented:
er, wait. that option is already there in the form of `--lifetime`

whyrusleeping commented:
@jbenet do you want `--lifetime` renamed to `--ttl`? or are we good to merge?

jbenet commented:
> @jbenet do you want --lifetime renamed to --ttl? or are we good to merge?

ah, no, good to merge.

jbenet commented:
no, dont merge yet. still CRing

jbenet commented on namesys/routing.go:
```diff
@@ -19,25 +22,55 @@ var log = logging.Logger("namesys")
+	cacheTil := time.Now().Add(r.cachelife)
+	eol, ok := checkEOL(rec)
+	if ok && eol.Before(cacheTil) {
+		cacheTil = eol
```
um... this isn't used anywhere.

whyrusleeping replied:
this is an interesting case of go's unused variable error. Its used, then assigned to, then that assigned value is never used.

jbenet commented on namesys/routing.go:
```diff
@@ -19,25 +22,55 @@ var log = logging.Logger("namesys")
+	r.cachelock.Lock()
+	r.cache[name] = cacheEntry{
+		val: val,
+		eol: eol,
```
was this supposed to be `cacheTil`?

whyrusleeping replied:
yeap... i love copy pasting things.

gamemanj commented:
I'm a bit worried this will hide speed problems enough that they don't get fixed...
Also, it makes the "IPNS livestream" idea kind of problematic, since you'll have a minute long delay on new updates thanks to the cache.

whyrusleeping commented:
@gamemanj read the comment chain.

whyrusleeping commented:
@jbenet actually used that variable.

ion1 commented:
`publish --lifetime=48hours --ttl=1hour` would be perfectly sensible:

* The record is considered invalid 48 hours after the *publish time*.
* You don’t need to ask the network for possible updates to the record for 1 hour after *having received a copy*, i.e. I’m okay with others keeping a stale version of my record for up to 1 hour.

The `Cache-Control/max-age` header is determined by `--ttl` (the semantics match perfectly), not `--lifetime`.

whyrusleeping commented:
@ion1 @gamemanj I also want to point out that live streaming will be done best by pubsub, ipns polling for streaming is equatable to setting your dns records to each frame of the video, one at a time. Not to say that you shouldnt try it, and that ipns *should* be slow, but thats just not its intended purpose.

whyrusleeping commented:
alright, theres a `--ttl` option.

jbenet commented on routing/dht/routing.go:
```diff
@@ -84,7 +84,10 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key key.Key) ([]byte, error) {
 
-	vals, err := dht.GetValues(ctx, key, 16)
+	// retrieve a majority of the expected record count
+	majority := (KValue / 2) + 1
```
again, this needs a real justification, else we should keep it at 16. the probabilities will seriously hurt when n < 10.

whyrusleeping replied:
what do you mean by n < 10? I'm not sure what youre trying to say

jbenet replied:
What is `KValue` and consequently what is `(KValue / 2) + 1`?

whyrusleeping replied:
20 and 11

jbenet replied:
The probability that a user will find a fresh record depends on things like: 
- the size of the network
- the churn of the network
- the freshness of the routing table
- the number of records sought.

and the number of records is (or is roughly like) an exponent on the probability of NOT finding the freshest record. the take away is that a _very small_ number (<10) is not going to give good results in the presence of (even modest) churn, or a not-perfectly-formed routing table. rising above `10` significantly improves the probability.

--- 

please do not change numbers if you do not fully understand the algorithmic implications.

there are other ways to speed up the resolution rather than "not looking". instead we should be:

- cutting down the time for dials
- _ensuring_ that the resolution is only one way so it terminates quickly, if there wont be 16 records. (the more i look at the dht the more i'm convinced it's exhausting the network, not just the subset of the network in the gradient (towards the lowest XOR distance) of the key).

i suspect the slowness here is that the node finds all the records available on the network but still searches through peers that are _not_ in the gradient.

whyrusleeping replied:
well, looks like i had the wrong number in my head for K, its 10, not 20. That explains a lot.

jbenet replied:
> 20 and 11.

No, `KValue = 10`. `(KValue / 2) + 1 = 6`.

we maybe should use `KValue` for this, but it would have to be the whole `KValue` and `KValue` should then increase (likely not what we want). we can instead make another parameter:

```
const RecordLookupThreshold = 16
```

---

Also, this comment is incorrect.
```
// K is the maximum number of requests to perform before returning failure.
```
is incorrect. if we want the `k` from kademlia, then it _should_ mean:

```
// K is the number of closest peers to find
```
and it looks like we're mostly using it this way here:

```
jbenet @ lorien : ~/go/src/github.com/ipfs/go-ipfs/routing/dht * master > grep KValue *
handlers.go:var CloserPeerCount = KValue
lookup.go:	tablepeers := dht.routingTable.NearestPeers(kb.ConvertKey(key), KValue)
lookup.go:	out := make(chan peer.ID, KValue)
lookup.go:	peerset := pset.NewLimited(KValue)
grep: pb: Is a directory
routing.go:	rtp := dht.routingTable.NearestPeers(kb.ConvertKey(key), KValue)
routing.go:	for p := range dht.FindProvidersAsync(ctx, key, KValue) {
routing.go:	peers := dht.routingTable.NearestPeers(kb.ConvertKey(key), KValue)
routing.go:	peers := dht.routingTable.NearestPeers(kb.ConvertPeerID(id), KValue)
routing.go:	peers := dht.routingTable.NearestPeers(kb.ConvertPeerID(id), KValue)
util.go:var KValue = 10
```

whyrusleeping replied:
Alright, so heres the part I disagree with. I dont think the number of records we require should be greater than the number of records that a single store operation puts. That is the reason for a lot of the slowness, we put out 10 records, and then try to get 16 from the network.

whyrusleeping replied:
I'm confident that we are fine lowering the value from 16 to 11, every time a query is made, if a node requests K, and receives values A B B A C B, and A is the most valid record, then the node making the query will 'correct' all the nodes who did not send A by sending them a STORE rpc with the value A. Because of this, the probabilities of getting the freshest record are greatly improved.

All of this is assuming that we change K from 10 to 20 (as recommended by the spec, which is also the reason I assumed it was 20 this whole time)

jbenet replied:
> I dont think the number of records we require should be greater than the number of records that a single store operation puts.

Yep, agreed, feel free to bump that up!

> That is the reason for a lot of the slowness, we put out 10 records, and then try to get 16 from the network.

No, the slowness is that the DHT currently _does not_ exhaust the query set correctly-- it considers nodes NOT on the gradient path. It ends up talking to way more nodes than it should before terminating a query.

Do you understand what i'm saying here? o/ 

> if a node requests K, and receives values A B B A C B, and A is the most valid record, then the node making the query will 'correct' all the nodes who did not send A by sending them a STORE rpc with the value A. Because of this, the probabilities of getting the freshest record are greatly improved.

of course, that is a very useful mechanic. However it does not _at all_ guarantee that -- after natural churn in the network -- nodes will actually be able to find the freshest records in the first place! And this wont be immediately apparent, it will be an insidious bug that harms people over time but is tricky to reproduce. 

For IPNS, _**Safety**_ is critical.  _**Liveness**_.  is achieved through optimization, but never at the cost of safety. 

---

please go back and reread my comments -- sp https://github.com/ipfs/go-ipfs/pull/1887/files#r43575580 --, and reread the relevant papers (like s/kademlia http://www.tm.uka.de/doc/SKademlia_2007.pdf (the query resolution parts) and the survey papers on DHTs). this is not an argument that is useful to have without a strong grounding in the actual probabilistic models these types of systems assume, and trying to explain it in this issue is not going to do even a tenth of a job as good as those papers.

ion1 commented:
It might be good to have code to clean up old entries from the cache, this could be a slight memory leak otherwise.

There seems to be a LRU cache implementation here: https://github.com/hashicorp/golang-lru

whyrusleeping commented:
changed the value back to 16, and changed the cache to an lru, with a configurable size.

No other DHT parameters changed here, weird things started happening when i tweaked K, so i'm not going to bother until later.

ion1 commented:
Nice. Is the cache lifetime limit useful in addition to the LRU policy though? How about letting popular records with long TTLs be cached as long as their TTL value allows?

You could get the TTL value from the record, fall back to a default of one minute, compute `time.Now().Add(ttl)`, limit that according to the record EOL and use that as the cache expiry for that record. I could then return the cache entry’s `expiry.Sub(time.Now())` as the current TTL in https://github.com/ipfs/go-ipfs/issues/1921.

ion1 commented:
Also, the LRU cache implementation does its own locking.

cryptix commented on core/commands/publish.go:
```diff
@@ -96,7 +97,18 @@ Publish an <ipfs-path> to another public key (not implemented):
+				return
+			}
+
+			ctx = context.WithValue(ctx, "ipns-publish-ttl", d)
```
:+1: we should enhance the context much more often.

This would allow us to get a trace of where it failed in timeouts instead of a blank "deadline exceeded".

cryptix commented on namesys/routing.go:
```diff
@@ -19,25 +23,74 @@ var log = logging.Logger("namesys")
-		panic("attempt to create resolver with nil routing system")
+func (r *routingResolver) cacheGet(name string) (path.Path, bool) {
+	r.cachelock.Lock()
+	ientry, ok := r.cache.Get(name)
```
2nding @ion1 here. The locking isn't needed. 

https://github.com/hashicorp/golang-lru/blob/master/lru.go#L87

cryptix commented on namesys/routing.go:
```diff
@@ -19,25 +23,74 @@ var log = logging.Logger("namesys")
+		val: val,
+		eol: cacheTil,
+	})
+	r.cachelock.Unlock()
```
again, not needed.

cryptix commented on namesys/routing.go:
```diff
@@ -19,25 +23,74 @@ var log = logging.Logger("namesys")
+
+	entry, ok := ientry.(cacheEntry)
+	if !ok {
+		return "", false
```
The cache get `ok` is shadowed here.

Also: there is no differentiation here between a malformed and an elapsed entry. Could you insert a warning print?

cryptix commented on namesys/namesys.go:
```diff
@@ -26,12 +26,12 @@ type mpns struct {
 }
 
 // NewNameSystem will construct the IPFS naming system based on Routing
```
Maybe update the comment with sensible defaults?

ion1 commented:
@whyrusleeping, please see my commits in https://github.com/ion1/go-ipfs/commits/feat/ipns-cache

The first commit has some fixes for things including the redundant locking and `lru.New` failing given a zero size.

The second commit removes cachelife with the assumption that the LRU policy along with respecting the TTL (and EOL) fields is enough.

cryptix commented:
@whyrusleeping I also fixed your tests.. :) Qmduc5TqSk5hWnuvBCDHQMXnPALPZ131Jz9wbU9SYTbjyB

whyrusleeping commented:
@jbenet this is RFM i beleive. @ion1 @cryptix can i get some thumbs ups?

cryptix commented:
:+1: used it already

ion1 commented:
https://github.com/ion1/go-ipfs/commit/f531ad3c65a67960effad9cadea4fcdd18a519a1

I refactored the EOL computation out of `cacheSet`. I’m going to need its result for the Cache-Control PR, it would be a bit hairy to pull the value out of `cacheSet`.

ion1 commented:
https://github.com/ion1/go-ipfs/commit/bd8489473fd3d9ec2afafc5c6f8820583372dbcb

A bit better: return both the absolute `eol` and the relative `ttl` and use `ttl > 0` to decide whether the result should be cached. This will also make it nicer for the Cache-Control PR.

ion1 commented:
```
<whyrusleeping> ion: okay. Just go ahead and apply that in your PR once 1887 merges
```

Alright.

:+1:, especially after squashing so one commit doesn’t add cachelife just for another to remove it.

jbenet commented:
this LGTM

jbenet commented:
@whyrusleeping @ion1 @cryptix damn i merged this without realizing-- there's no tests. can we write some?

whyrusleeping commented:
@jbenet sure thing! https://github.com/ipfs/go-ipfs/issues/1934
